File: code/processing/seq/20210908_lacI_titration/barcode_extraction.jl
using FASTX, PyCall, LightXML, CodecZlib, BioSequences, DataFrames, CSV, Distributed
@everywhere begin 
    using FASTX, PyCall, LightXML, CodecZlib, BioSequences, DataFrames, CSV

    using PyCall
    git = pyimport("git")
    glob = pyimport("glob")

    function get_seq(record)
        return identifier(record), sequence(record)
    end
    
    # Define sequence next to GFP barcode
    ref_seq = LongDNASeq("TTATTTGTACAGTT")

    function find_gfp_bc(seq; ref_seq=ref_seq, ind_filter=nothing)

        ind = findfirst(ref_seq, seq)
        if isnothing(ind)
            gfp_idx = missing
            gfp_bc = missing
        else
            gfp_idx = ind[1]-5
            if ~isnothing(ind_filter)
                if gfp_idx != ind_filter
                    gfp_idx = missing
                    gfp_bc = missing
                else
                    gfp_bc = seq[ind[1]-5:ind[1]-1]
                end
            else
                gfp_bc = seq[ind[1]-5:ind[1]-1]
            end     
        end

        return gfp_idx, gfp_bc
    end

    # Date for sequencing run of the library to map
    DATE = 20210908
    # Description to be attached to folder names
    DESCRIPTION = "_lacI_titration/"

    # Find home directory for repo
    repo = git.Repo("./", search_parent_directories=true)
    homedir = repo.working_dir
    processed_dir = homedir * "/data/processed_sequencing/$(DATE)$(DESCRIPTION)"
    datadir = homedir * "/data/raw_sequencing/$(DATE)$(DESCRIPTION)"

    outputdir = homedir * "/data/barcodes/$(DATE)$(DESCRIPTION)"
    if ~isdir(outputdir)
        mkdir(outputdir)
    end

    # Find sequencing files
    files = glob.glob(processed_dir * "*fastq.gz")

end
Threads.@threads for file in files
    println("Reading file $file...")
    r = FASTQ.Reader(GzipDecompressorStream(open(file)))
    seq_list = map(x -> get_seq(x), collect(r))#[1:n_samples])
    
    println("Extracting barcodes...")
    # Initialize dataframe to save sequences
    Names = [:id, :seq]
    df_seq = DataFrame(map(idx -> getindex.(seq_list, idx), eachindex(first(seq_list))), Names)

    # Add index and sequence length to dataframe
    df_seq[!, "seq_len"] = length.(seq_list)

    # Extract barcode
    df_seq[!, "op_bc"] = map(x->x[1:20], df_seq[!, "seq"])

    println("Extracting GFP barcodes...")
    df_seq[!, "gfp_bc"] = map(x->find_gfp_bc(x, ind_filter=27)[2], df_seq[!, "seq"])
    df_seq = dropmissing(df_seq, :gfp_bc)


    df_gfp = CSV.read("./gfp_barcode.csv", DataFrame)

    # Define if barcode is present
    bc_bool = [String(x) in df_gfp[!, "gfp_barcode_revcomp"] for x in df_seq[!, "gfp_bc"]]
    df_seq_filt = df_seq[bc_bool, :]
    
    println("Writing file...")
    output_file_name = split(split(file, "/")[end], ".")[1]
    CSV.write(outputdir * output_file_name * ".csv", df_seq_filt)
end

File: code/processing/seq/20210908_lacI_titration/barcode_extraction.py
# %%
import os
import glob
import itertools
import re
import regex
import numpy as np
import pandas as pd
import skbio
import collections
import git
import multiprocessing
import rnaseq_barcode as rnaseq
import sys


# Check if number of cores is given
if len(sys.argv) > 1:
    cores = int(sys.argv[1])
else:
    cores = 1

# Find home directory for repo
repo = git.Repo("./", search_parent_directories=True)
homedir = repo.working_dir

# Date for sequencing run of the library to map
DATE = 20210908
# Description to be attached to folder names
DESCRIPTION = "_lacI_titration/"

# Find home directory for repo
repo = git.Repo("./", search_parent_directories=True)
homedir = repo.working_dir
datadir = homedir + f'/data/processed_sequencing/{DATE}{DESCRIPTION}'

# Find sequencing files
files = glob.glob(datadir + "*.fastq.gz")

# output directory
outdir = f"{homedir}/data/barcodes/{DATE}{DESCRIPTION}"

if not os.path.exists(outdir):
    os.makedirs(outdir)
    
# read reference GFP barcodes
df_gfp = pd.read_csv("./gfp_barcode.csv")


def find_close_bc(seq, barcodes):
    bc_lists = [list(x) for x in barcodes]
    for j, bc in enumerate(bc_lists):
        if np.sum([x != y for x, y in zip(list(seq), bc)]) == 1:
            return barcodes[j]
        
    return np.nan


# %%
def analyze_file(file):
    print("Reading sequences from fastq file")
    df_seq = rnaseq.seq.read_sequences(file)

    # Add index and sequence length to dataframe
    df_seq["index"] = file.split("/")[-1]
    df_seq["seq_len"] = df_seq.sequence.apply(len)

    print("Mapping operator and GFP barcode")
    # Extract barcode
    df_seq = df_seq.assign(op_bc=[x[0:20] for x in df_seq.sequence])

    # Define sequence next to GFP barcode
    ref_seq = "TTATTTGTACAGTT"

    # Find GFP barcodes and filter for correct location
    gfp_bc_list = df_seq['sequence'].apply(rnaseq.seq.find_bc, args=(ref_seq, 26, 1))
    df_seq['gfp_idc'], df_seq['gfp_bc'] = zip(*gfp_bc_list)
    df_seq.head()
    
    print("Filtering GFP barcodes")
    df_filt = df_seq.dropna(subset=['gfp_bc'])
    bc_lists = [list(x) for x in df_gfp.gfp_barcode_revcomp.values]

    # Define if barcode is present
    bc_bool = [x not in df_gfp.gfp_barcode_revcomp.values for x in df_filt.gfp_bc.values]

    df_filt.loc[bc_bool,'gfp_bc'] = df_filt.loc[bc_bool,'gfp_bc'].apply(find_close_bc, args=(df_gfp.gfp_barcode_revcomp.values,))
    df_filt = df_filt.dropna(subset=['gfp_bc'])

    # Keep only desired columns
    df_filt = df_filt[["sequence", "index", "op_bc", "gfp_bc"]]
    print("Saving into memory")

    file_name = file.split("/")[-1].replace('.fastq.gz', '.csv')
    df_filt.to_csv(f"{outdir}{file_name}", index=False)

    
# Parallelize analysis    
a_pool = multiprocessing.Pool(cores)
result = a_pool.map(analyze_file, files)
